morpheus/_lib/src/io/loaders/rest.cpp

```diff
@@ -17,21 +17,238 @@
 
 #include "morpheus/io/loaders/rest.hpp"
 
+#include <boost/algorithm/string.hpp>
+#include <boost/asio.hpp>
+#include <boost/beast/core.hpp>
+#include <boost/beast/http.hpp>
+#include <boost/beast/version.hpp>
 #include <glog/logging.h>
 #include <nlohmann/json.hpp>
+#include <pybind11/cast.h>
+#include <pybind11/gil.h>
+#include <pybind11/pybind11.h>
+#include <pybind11/pytypes.h>
+#include <pymrc/utilities/object_cache.hpp>
 
 #include <memory>
 #include <ostream>
 #include <stdexcept>
 
+namespace py = pybind11;
+
+namespace beast = boost::beast;
+namespace http  = beast::http;
+namespace net   = boost::asio;
+using tcp       = net::ip::tcp;
+
+#define PORT "80"
+
 namespace morpheus {
 RESTDataLoader::RESTDataLoader(nlohmann::json config) : Loader(config) {}
 
+void get_data_from_endpoint(http::response<http::dynamic_body>& response,
+                            const std::string& method,
+                            const std::string& endpoint,
+                            const std::string& params,
+                            const std::string& content_type,
+                            const std::string& body,
+                            const std::unordered_map<std::string, std::string>& x_headers,
+                            int max_retry)
+{
+    py::gil_scoped_acquire gil;
+    py::module_ urllib = py::module::import("urllib.parse");
+    std::string ep(endpoint);
+    py::object ep_pystr = py::str(ep);
+
+    // Following the syntax specifications in RFC 1808, urlparse recognizes a netloc only if it is properly introduced
+    // by ‘//’. Otherwise the input is presumed to be a relative URL and thus to start with a path component. 
+    // ref: https://docs.python.org/3/library/urllib.parse.html
+    // Workaround here is to prepend "//" if endpoint does not already include one
+    if (ep_pystr.attr("find")("//").cast<int>() == -1)
+    {
+        ep = "//" + ep;
+    }
+
+    py::object result  = urllib.attr("urlparse")(ep);
+    std::string host   = result.attr("hostname").is_none() ? "" : result.attr("hostname").cast<std::string>();
+    std::string target = result.attr("path").is_none() ? "" : result.attr("path").cast<std::string>();
+    if (target.empty())
+    {
+        target = "/";
+    }
+    std::string query = result.attr("query").is_none() ? "" : result.attr("query").cast<std::string>();
+    // if params exists, overrides query part included in endpoint
+    if (!params.empty())
+    {
+        query = params;
+    }
+    py::gil_scoped_release gil_release;
+
+    net::io_context ioc;
+    tcp::resolver resolver(ioc);
+    beast::tcp_stream stream(ioc);
+    try
+    {
+        auto const endpoint_results = resolver.resolve(host, PORT);
```

Comment:
I'd like to see this made a bit more declarative, it improves readability and makes updating things easier in the future.

For example:

```c++
auto verb = get_http_verb(method);
auto request = construct_request(verb, target, query, count);

init_request(request);

try_get_data(request, response, max_retry, interval, ...);
```